### PR TITLE
[Bugfix][ROCm] Reject incompatible mixed KV-cache layouts

### LINKED ISSUE #50237: [Bug]: ROCm mixed KV-cache layouts abort speculative decoding with AssertionError
### Your current environment

<details>
<summary>Relevant environment</summary>

```text
vLLM: 0.26.0 ROCm image (also present on main at 5b14019)
PyTorch: 2.11.0+gitd0c8b1f
ROCm/HIP: 7.2.53211
Python: 3.12.13
GPU host: 8x AMD Instinct MI308X (gfx942)
```

</details>

### Describe the bug

On ROCm, a blocks-first target backend and K/V-first speculative-draft backend
can share one KV allocation. `_align_mixed_attention_kv_cache_views` then applies
the encoder-decoder restride path to both layouts. A fused blocks-first cache such
as MiniMax-M3's `(num_blocks, block_size, num_heads, 2 * head_size)` has no
explicit K/V dimension at index 1, so startup aborts at:

```text
vllm/v1/worker/gpu/attn_utils.py:444
assert kv_cache.shape[1] == 2
AssertionError
```

Weight-free reproduction:

```python
import torch
from vllm.v1.worker.gpu.attn_utils import (
    _restride_blocks_first_kv_cache_to_kv_first_storage,
)

kv_cache = torch.zeros(3, 16, 1, 4)
_restride_blocks_first_kv_cache_to_kv_first_storage(kv_cache)
```

Expected: reject incompatible layouts with an actionable backend-configuration
error. Current workaround: set `attention_backend` explicitly in the speculative
config so target and drafter use matching layouts.

I searched open and closed issues/PRs. #47882 is related backend-selection work
for an EAGLE3 performance case, but does not cover mixed-layout validation or this
startup assertion.

### PR DESCRIPTION
**Pull request by 2imi9**
## Purpose

Fixes #50237.

Mixed block dimensions are only safe for the encoder-decoder cross-attention
case this alignment path was designed for. A speculative target/drafter pair can
otherwise hit an assertion, or bypass the shape check when it has two KV heads.
Validate the blocks-first layer is cross-attention before restriding and report
the layer, shape, and `speculative_config.attention_backend` remedy. Backend
selection is unchanged.

AI assistance: OpenAI Codex assisted with investigation and drafting. The human
submitter reviewed and understands every changed line.

## Test Plan

```text
.venv/bin/pre-commit run --files vllm/v1/worker/gpu/attn_utils.py tests/v1/worker/test_attn_utils.py
.venv/bin/pre-commit run --hook-stage manual mypy-3.12 --files vllm/v1/worker/gpu/attn_utils.py tests/v1/worker/test_attn_utils.py
pytest tests/v1/worker/test_attn_utils.py -q
```

Run stock, patched, and matching-backend startup arms on 8x MI308X using the
same vLLM 0.26.0 ROCm image and MiniMax-M3/DSpark inputs.

## Test Result

- Contributor hooks and Python 3.12 mypy pass.
- The complete unit file passes: `6 passed`. With two KV heads, the new negative
  test fails against stock because the incompatible layout does not raise.
- Stock target `TRITON_ATTN` plus drafter `ROCM_ATTN` reproduces the assertion on
  all workers.
- The final patched arm reports the non-cross-attention layer, shape
  `(213012, 1, 128, 256)`, and the matching-backend remedy.
- Stock target and drafter `TRITON_ATTN` reach readiness; all four smoke prompts
  finish with `stop` (230/360 draft tokens accepted).

No benchmark or LMCache workload was run.

### CODE REVIEW

**Comment by 2imi9** [comment]
Rebased onto current main (0f3a7cf -> d569034) to clear the out-of-date status; DCO and mergify are green on the new head. `pre-commit / pre-run-check` fails only on the contributor gate (this account has 1 merged PR, below the 4 required). Could a maintainer add the `ready` label so pre-commit and Buildkite CI can run? Pre-commit and mypy 3.12 pass locally on the rebased head; test details are in the PR description.
